lib/autoupdate/rolloutcontroller/client_test.go

```diff
@@ -0,0 +1,188 @@
+	require.Equal(t, len(stubs.updateRolloutAnswers), len(stubs.updateRolloutExpects), "invalid stubs, update validations and answers slices are not the same length")
+
+	return &mockClient{
+		getAutoUpdateConfig:          &getHandler[*autoupdate.AutoUpdateConfig]{t, stubs.configAnswers},
+		getAutoUpdateVersion:         &getHandler[*autoupdate.AutoUpdateVersion]{t, stubs.versionAnswers},
+		getAutoUpdateAgentRollout:    &getHandler[*autoupdate.AutoUpdateAgentRollout]{t, stubs.rolloutAnswers},
+		createAutoUpdateAgentRollout: &createUpdateHandler[*autoupdate.AutoUpdateAgentRollout]{t, stubs.createRolloutExpects, stubs.createRolloutAnswers},
+		updateAutoUpdateAgentRollout: &createUpdateHandler[*autoupdate.AutoUpdateAgentRollout]{t, stubs.updateRolloutExpects, stubs.updateRolloutAnswers},
+		deleteAutoUpdateAgentRollout: &deleteHandler{t, stubs.deleteRolloutAnswers},
+	}
+}
+
+type mockClientStubs struct {
+	configAnswers        []callAnswer[*autoupdate.AutoUpdateConfig]
+	versionAnswers       []callAnswer[*autoupdate.AutoUpdateVersion]
+	rolloutAnswers       []callAnswer[*autoupdate.AutoUpdateAgentRollout]
+	createRolloutAnswers []callAnswer[*autoupdate.AutoUpdateAgentRollout]
+	createRolloutExpects []require.ValueAssertionFunc
+	updateRolloutAnswers []callAnswer[*autoupdate.AutoUpdateAgentRollout]
+	updateRolloutExpects []require.ValueAssertionFunc
+	deleteRolloutAnswers []error
+}
+
+type callAnswer[T any] struct {
+	result T
+	err    error
+}
+
+// getHandler is used in a mock client to answer get resource requests during tests.
+// It takes a list of answers and errors and will return them when invoked.
+// If there are no stubs left it fails the test.
+type getHandler[T protoadapt.MessageV1] struct {
+	t       *testing.T
+	answers []callAnswer[T]
+}
+
+func (h *getHandler[T]) handle(_ context.Context) (T, error) {
+	if len(h.answers) == 0 {
+		require.Fail(h.t, "no answers left")
+	}
+
+	entry := h.answers[0]
+	h.answers = h.answers[1:]
+
+	// We need to deep copy because the reconciler might do updates in place.
+	// We don't want the original resource to be edited as this would mess with other tests.
+	return utils.CloneProtoMsg(entry.result), entry.err
+}
+
+// isEmpty returns true only if all stubs were consumed
+func (h *getHandler[T]) isEmpty() bool {
+	return len(h.answers) == 0
+}
+
+// createUpdateHandler is used in a mock client to answer create or update resource requests during tests (any request whose arity is 2).
+// It first validates the input suing the passed validation function, then it returns the predefined answer and error.
+// If there are no stubs left it fails the test.
+type createUpdateHandler[T protoadapt.MessageV1] struct {
+	t       *testing.T
+	expect  []require.ValueAssertionFunc
+	answers []callAnswer[T]
+}
+
+func (h *createUpdateHandler[T]) handle(_ context.Context, object T) (T, error) {
+	if len(h.expect) == 0 {
+		require.Fail(h.t, "not expecting more calls")
```

Comment:
Is this guaranteed to be executed in the goroutine running the test? The `require` helpers will fail the test and then goexit, which won't actually stop the test if it's a goroutine calling the helpers.

Reply:
Currently yes, and I don't plan to write multi-goroutine tests if possible. Single-routine tests are easier to await and debug. I'm losing a lot of time because of the async reconciliation in the operator tests and prefer to do everything synchronously + use require.